ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/StateTreeManager.java

```diff
@@ -84,37 +74,62 @@ public void onValidatorRegistered(ValidatorRegistrationEvent event) {
 
   @Subscribe
   public void onNewSlot(Date date) {
-    requireNonNull(state);
     LOG.info("****** New Slot at: " + date + " ******");
-    Optional<BeaconBlock> block = this.storage.getUnprocessedBlock();
+    Optional<BeaconBlock> block = this.store.getUnprocessedBlock();
     if (block.isPresent()) {
-      LOG.info("Unprocessed block retrieved.");
+      Bytes32 blockStateRoot = block.get().getState_root();
       try {
-        // get state corresponding to the block
-        this.state = BeaconState.deepCopy(state);
-        stateTransition.initiate(this.state, block.get());
-        this.storage.addProcessedBlock(block.get().getState_root(), block.get());
-      } catch (NoSuchElementException e) {
-        LOG.warn(e.toString());
+        Optional<BeaconBlock> parentBlock = this.store.getParent(block.get());
+        Bytes32 blockRoot = HashTreeUtil.hash_tree_root(block.get().toBytes());
+        if (parentBlock.isPresent()) {
+          // get state corresponding to the parent block
+          Optional<BeaconState> state = this.store.getState(parentBlock.get().getState_root());
+          BeaconState newState = BeaconState.deepCopy(state.get());
+          stateTransition.initiate(newState, block.get());
+          Bytes32 stateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
+          LOG.info("New state root: " + stateRoot.toHexString());
+          // state root verification
+          checkArgument(
+              blockStateRoot.equals(stateRoot),
+              "The block's state root %s does not match the calculated state root %s!",
+              blockStateRoot.toHexString(),
+              stateRoot.toHexString());
+          // TODO: storing block and state together as a tuple would be more convenient
+          this.store.addState(stateRoot, newState);
+          this.store.addProcessedBlock(blockStateRoot, block.get());
+          this.store.addProcessedBlock(blockRoot, block.get());
```

Comment:
Are these lines duplicate? Do they not mean the same thing?

Reply:
No

Reply:
I mean, I’m not sure if I understand why we add the same block to the store twice.

Reply:
No, but it looks weird. I needed to lookup a block by blockroot and by state.  This will be done in a better way going forward

Reply:
Sorry, trying to answer on my phone...not working well. I need to be able to lookup a block by state root AND by parent block root. Therefore I need it stored twice. I will optimize layer